File: js/channel.ts
/**
 * Z-Wing post-handshake encrypted channel for TypeScript.
 *
 * Wraps a `DuplexStream` with ChaCha20-Poly1305 records using
 * sequence-numbered nonces. Wire format matches the Go and Rust
 * implementations byte-for-byte.
 */

import { chacha20poly1305 } from "@noble/ciphers/chacha.js";

import {
  DuplexStream,
  HandshakeOutput,
  MAX_FRAME_SIZE,
  readFrame,
  writeFrame,
} from "./handshake.js";
import type { IdentityPublic } from "./identity.js";

const CHACHA_OVERHEAD = 16;
const NONCE_LEN = 12;

/** Build the 12-byte nonce from a 64-bit BE counter. */
export function nonceFor(seq: bigint): Uint8Array {
  const n = new Uint8Array(NONCE_LEN);
  for (let i = 0; i < 8; i++) {
    n[NONCE_LEN - 1 - i] = Number((seq >> BigInt(i * 8)) & 0xffn);
  }
  return n;
}

export class Channel {
  private inner: DuplexStream;
  public readonly remote: IdentityPublic;
  private rxKey: Uint8Array;
  private txKey: Uint8Array;
  private rxSeq = 0n;
  private txSeq = 0n;
  private rxOverflow: Uint8Array | null = null;

  constructor(inner: DuplexStream, out: HandshakeOutput, initiator: boolean) {
    this.inner = inner;
    this.remote = out.remote;
    if (initiator) {
      this.txKey = out.keyI2R;
      this.rxKey = out.keyR2I;
    } else {
      this.txKey = out.keyR2I;
      this.rxKey = out.keyI2R;
    }
    // Best-effort wipe of the handshake-side keys.
    out.shared.fill(0);
    // Note: keyI2R / keyR2I have been copied into rxKey / txKey above
    // by reference — wiping them is the responsibility of the caller
    // since we don't own the original buffers exclusively.
  }

  /** Encrypt and send a single record. Splits if larger than MAX. */
  async send(plaintext: Uint8Array): Promise<void> {
    const maxPlain = MAX_FRAME_SIZE - CHACHA_OVERHEAD;
    let written = 0;
    while (written < plaintext.length) {
      const end = Math.min(written + maxPlain, plaintext.length);
      const chunk = plaintext.slice(written, end);
      if (this.txSeq === 0xffffffffffffffffn) {
        throw new Error("zwing: AEAD sequence number exhausted");
      }
      const nonce = nonceFor(this.txSeq);
      const ct = chacha20poly1305(this.txKey, nonce).encrypt(chunk);
      this.txSeq += 1n;
      await writeFrame(this.inner, ct);
      written = end;
    }
  }

  /** Receive one full record. */
  async recv(): Promise<Uint8Array> {
    if (this.rxOverflow) {
      const buf = this.rxOverflow;
      this.rxOverflow = null;
      return buf;
    }
    const frame = await readFrame(this.inner);
    if (this.rxSeq === 0xffffffffffffffffn) {
      throw new Error("zwing: AEAD sequence number exhausted");
    }
    const nonce = nonceFor(this.rxSeq);
    const pt = chacha20poly1305(this.rxKey, nonce).decrypt(frame);
    this.rxSeq += 1n;
    return pt;
  }
}

File: py/channel.py
"""Z-Wing post-handshake encrypted channel (Python).

ChaCha20-Poly1305 over length-prefixed frames with sequence-numbered
nonces. Wire format identical to Go and Rust.
"""

from __future__ import annotations

import struct
from typing import Optional

from cryptography.hazmat.primitives.ciphers.aead import ChaCha20Poly1305

from handshake import DuplexStream, HandshakeOutput, MAX_FRAME_SIZE, read_frame, write_frame
from identity import IdentityPublic

CHACHA_OVERHEAD = 16
NONCE_LEN = 12


def nonce_for(seq: int) -> bytes:
    """12-byte nonce: high 4 bytes zero, low 8 bytes BE counter."""
    n = bytearray(NONCE_LEN)
    n[NONCE_LEN - 8 :] = struct.pack(">Q", seq)
    return bytes(n)


class Channel:
    """Post-handshake Z-Wing secure channel."""

    def __init__(
        self, inner: DuplexStream, out: HandshakeOutput, initiator: bool
    ) -> None:
        self._inner = inner
        self.remote: IdentityPublic = out.remote
        if initiator:
            tx_key = out.key_i2r
            rx_key = out.key_r2i
        else:
            tx_key = out.key_r2i
            rx_key = out.key_i2r
        self._tx = ChaCha20Poly1305(tx_key)
        self._rx = ChaCha20Poly1305(rx_key)
        self._tx_seq = 0
        self._rx_seq = 0
        self._rx_overflow: Optional[bytes] = None

    def send(self, plaintext: bytes) -> None:
        max_plain = MAX_FRAME_SIZE - CHACHA_OVERHEAD
        written = 0
        while written < len(plaintext):
            end = min(written + max_plain, len(plaintext))
            chunk = plaintext[written:end]
            if self._tx_seq >= (1 << 64) - 1:
                raise ValueError("zwing: AEAD sequence number exhausted")
            nonce = nonce_for(self._tx_seq)
            ct = self._tx.encrypt(nonce, chunk, None)
            self._tx_seq += 1
            write_frame(self._inner, ct)
            written = end

    def recv(self) -> bytes:
        if self._rx_overflow is not None:
            buf = self._rx_overflow
            self._rx_overflow = None
            return buf
        frame = read_frame(self._inner)
        if self._rx_seq >= (1 << 64) - 1:
            raise ValueError("zwing: AEAD sequence number exhausted")
        nonce = nonce_for(self._rx_seq)
        pt = self._rx.decrypt(nonce, frame, None)
        self._rx_seq += 1
        return pt
